Cleanup Add-Type

iSazonov opened the pull request:
## PR Summary
Related #5725.

It's part of the big PR I'm trying to break into smaller pieces.
Subsequent PRs will focus on enhancements.

- Remove CSharp language versions and JScript from Language enum. Later we can use Roslyn parse option if we need set a language version. JScript is not in Roslyn. It is a breaking change (Approved by @SteveL-MSFT in #5725).
- Correct comments.
- Remove IgnoreWarnings parameter.  It is not implemented and no plans do that. We could use the standard parameter `WarningAction` if we'll need.

## PR Checklist

Note: Please mark anything not applicable to this PR `NA`.

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - [x] Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [NA] User facing [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - [NA] Issue filed - Issue link:
- [ ] [Change is not breaking](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)
- [x] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
    - [NA] [Add `[feature]` if the change is significant or affects feature tests](https://github.com/PowerShell/PowerShell/blob/master/docs/testing-guidelines/testing-guidelines.md#requesting-additional-tests-for-a-pr)
- [x] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` to the beginning of the title and remove the prefix when the PR is ready.

iSazonov committed:
Cleanup Add-Type

Remove CSharp language versions and JScript from Language enum.
Correct comments.
Remove IgnoreWarnings parameter.

iSazonov commented:
@SteveL-MSFT We need an approve from PowerShell Committee.

daxian-dbw commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/AddType.cs:
```diff
@@ -553,149 +509,110 @@ internal string GenerateTypeSource(string typeNamespace, string name, string sou
+                default:
+                    {
+                        Diagnostics.Assert(false, "GetUsingSet: Unsupported language family.");
+                    }
```
No need to put the assert in a code block `{}`
One more minor thing -- the assert is put in `default` case here while not in the above similar switch code, maybe you can keep them consistent.

iSazonov replied:
The PR is formal cleanup to reduce the size of subsequent PRs to simplify a code review (my develop branch has over 1200 changed lines and I still haven't done all the work) - latter we'll again edit the code. So I don't think it makes sense to make today such minor changes.

daxian-dbw replied:
If you are going to push more commits to the same PR, then just ignore the comments now.
If you want this PR to be merged before doing the further work, then I think you need to resolve those comments.

daxian-dbw replied:
> One more minor thing --

The minor part is not a blocking comment, you can leave it as is.

iSazonov replied:
About `Diagnostics.Assert` - code block removed.
About minor comment - here we should return value, above we use formal `return null;`

daxian-dbw replied:
Thanks for the update. I removed the extra spaces before `Diagnostics.Assert`

daxian-dbw commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/AddType.cs:
```diff
@@ -1106,31 +1023,10 @@ private void WriteTypes(Assembly assembly)
-            if (IsCSharp(Language))
+            if (Language == Language.CSharp)
             {
                 switch (Language)
```
No need to switch on `Language` anymore, since you already checked `Language == Language.CSharp`

daxian-dbw replied:
The same. If you are going to push more commits to the same PR, then ignore this comment for now. Otherwise, this comment needs to be resolved for this PR to be merged.

iSazonov replied:
`switch` removed.

I want keep the PR small. No plans add new code.

daxian-dbw reviewed (changes requested):
Left two comments

SteveL-MSFT commented:
@PowerShell/powershell-committee reviewed this:
- ok with removing jscript since it's not supported by Roslyn
- ask is to investigate if passing the different versions of csharp are supported today with current implementation; prefer to not break existing users who may be passing the csharp version even though we agree that most users are likely not caring about specific version of csharp

So pending outcome of second bullet, we can decide if it's appropriate to remove the csharp versions from the enum.  Leaving as `Review - Committee` pending investigation.

iSazonov committed:
Correct switches

iSazonov commented:
1. GitHub search results (without repetitions and `Add-Type clones`):

| Version | Script | C#
|- | - | -
|`CSharpVersion3` | 118 | 0
|`CSharpVersion2` | 1 | 0
|`CSharpVersion1` | 1 | 0

`CSharpVersion4`, `CSharpVersion5`, `CSharpVersion6`, `CSharpVersion7`, `CSharpVersion7` was added in 
PowerShell Core 6.0. Nobody still use it.

Note that this cmdlet is rarely used in compilation mode, and with these exact language settings it is much less.

2. I don't know scenarios where users should use lower C# version with Roslyn. Why would anyone want to block new versions or features in his _own_ code? 
 I rather expect that users will request to add support to latest Roslyn and C# versions. Currently we have to change the cmdlet to add new version explicitly but if we'll use native Roslyn types we get this automatically. (With .Net Core hosting model even more automatically).

3. @SteveL-MSFT suggested using dynamic parameters because Roslyn exposes different types for different languages. This allow us to expose exact language version strings in a parameter in any cases. I don't investigate but I guess we could to try emulate current `-Language` behavior if we could give up current `Language` enum type.

daxian-dbw committed:
Remove unneeded spaces

daxian-dbw commented:
More information for the committee:
### In Windows PowerShell 5.1
the enum `Microsoft.PowerShell.Commands.Language` has the following members: `CSharp`, `CSharpVersion3`, `CSharpVersion2`, `VisualBasic`, `JScript`.
Among them, `CSharpVersion2` requires Microsoft .NET Framework 2 to be installed and `CSharpVersion3` requires Microsoft .NET Framework 3.5 to be installed. Therefore, on PSv5.1 these 2 options don't work unless the user manually installs .NET 2 and .NET 3.5. See examples below:
```
PS:2> Add-Type -TypeDefinition $source -Language CSharpVersion2
Add-Type : Cannot add type due to the following exception: Compiler executable file csc.exe cannot be found.. Verify
that Microsoft .NET Framework 2  is installed. On x64-based versions of Windows, you must also install the WOW64
feature.
At line:1 char:1
+ Add-Type -TypeDefinition $source -Language CSharpVersion2
+ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : NotSpecified: (:) [Add-Type], InvalidOperationException
    + FullyQualifiedErrorId : System.InvalidOperationException,Microsoft.PowerShell.Commands.AddTypeCommand

PS:3> Add-Type -TypeDefinition $source -Language CSharpVersion3
Add-Type : Cannot add type due to the following exception: Compiler executable file csc.exe cannot be found.. Verify
that Microsoft .NET Framework 3.5  is installed. On x64-based versions of Windows, you must also install the WOW64
feature.
At line:1 char:1
+ Add-Type -TypeDefinition $source -Language CSharpVersion3
+ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : NotSpecified: (:) [Add-Type], InvalidOperationException
    + FullyQualifiedErrorId : System.InvalidOperationException,Microsoft.PowerShell.Commands.AddTypeCommand
```

### In PowerShell Core
The following members were added to the enum `Language` when refactoring the code with Roslyn: `CSharpVersion1`, `CSharpVersion4`, `CSharpVersion5`, `CSharpVersion6`, `CSharpVersion7`.
Based on the search results from GitHub, none of them are used in any PowerShell scripts except in our tests.

### Summary

Given that `CSharpVersion2` and `CSharpVersion3` by default don't work in PSv5.1 (maybe PSv5 too), maybe it's not a too bad breaking change to remove them from `Language`. As for the other `CSharpVersionN` members, none of them are used anywhere.

iSazonov commented:
@SteveL-MSFT I'm really looking forward 😄

SteveL-MSFT commented:
@PowerShell/powershell-committee reviewed this and would be fine with removing the specific versions of csharp from the enum; we do not believe there is utility to having the different versions of the csharp language specified and for users that have specified a version they will get an error indicating to just use `csharp`.  

We will be reviewing the `Breaking Change Contract` and updating as necessary to reflect how we came to this decision.

daxian-dbw committed:
Add back the parameter "IgnoreWarnings"

The parameter `-IgnoreWarnings` is implemented in Windows PowerShell. It should also be implemented in PowerShell Core.

daxian-dbw commented:
The `-IgnoreWarnings` parameter shouldn't be removed yet. It's implemented in Windows PowerShell by setting [`CompilerParameters.TreatWarningsAsErrors = !((bool)ignoreWarnings)`](https://github.com/PowerShell/PowerShell/blob/v6.0.0-beta.1/src/Microsoft.PowerShell.Commands.Utility/commands/utility/AddType.cs#L1723).

The parameter was not implemented when porting to Roslyn in PowerShell Core, but that doesn't mean it should be removed from PowerShell Core. I think it makes sense to support it in PowerShell Core.

I pushed a commit to add this parameter back. Work can be done in future to actually support it in PowerShell Core.

iSazonov commented:
@daxian-dbw I agree with `-IgnoreWarnings`. I removed it because the name does not reflect the meaning - treat warnings as errors. I plan to add new parameter with name like `-WarnAsErrors` - we could add `IgnoreWarnings` as alias. 
Also I plan add support for preprocessor constants. The rest does not look so useful - if you have other considerations please let me know.

daxian-dbw commented:
> I plan to add new parameter with name like -WarnAsErrors - we could add IgnoreWarnings as alias.

That idea would work too :) The bottom line is that we need to continue to support this parameter in PowerShell Core. @iSazonov, thanks for initiating the rework of `Add-Type`!
